### ISSUE TITLE
Filtering Custom Expression from aggregated result fails with column not found

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#12839]
**Describe the bug**
It is not possible to filter a Custom Expression from aggregated result.

**To Reproduce**
1. Custom question > Sample Dataset > Products
2. Summarize "Average of Price" and Custom Expression `CountIf(interval([Created At], -1, "year"))` as "Last year" by Category
3. Filter the aggregated result by "Last year" greater than `0`:
![image](https://user-images.githubusercontent.com/1447303/86512573-a6bffa00-be03-11ea-8f23-81d07a119ce9.png)
4. Fails with `Column "source.Last year" not found; ...` (the error depends on the database queried, so might be `Unknown column 'source.Last year' in 'where clause'` or `Name Last year not found inside source at [2:2086]` for other database types)
![image](https://user-images.githubusercontent.com/1447303/86512585-b8090680-be03-11ea-9701-47a24616d993.png)
<details><summary>stacktrace</summary>

```
07-04 14:35:54 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2020-07-04T14:35:53.906530+02:00[Europe/Copenhagen]",
 :state "42S22",
 :json_query
 {:database 4,
  :query
  {:source-query
   {:source-table 10,
    :aggregation [["avg" ["field-id" 104]] ["aggregation-options" ["count-where" ["time-interval" ["field-id" 108] -1 "year"]] {:display-name "Last year"}]],
    :breakout [["field-id" 107]]},
   :filter [">" ["field-literal" "Last year" "type/Float"] 0]},
  :type "query",
  :parameters [],
  :middleware {:add-default-userland-constraints? true}},
 :native
 {:query
  "SELECT \"source\".\"CATEGORY\" AS \"CATEGORY\", \"source\".\"avg\" AS \"avg\", \"source\".\"count-where\" AS \"count-where\" FROM (SELECT \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" AS \"CATEGORY\", avg(\"PUBLIC\".\"PRODUCTS\".\"PRICE\") AS \"avg\", sum(CASE WHEN parsedatetime(year(\"PUBLIC\".\"PRODUCTS\".\"CREATED_AT\"), 'yyyy') = parsedatetime(year(dateadd('year', -1, now())), 'yyyy') THEN 1 ELSE 0.0 END) AS \"count-where\" FROM \"PUBLIC\".\"PRODUCTS\" GROUP BY \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" ORDER BY \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" ASC) \"source\" WHERE \"source\".\"Last year\" > 0 LIMIT 2000",
  :params nil},
 :status :failed,
 :class org.h2.jdbc.JdbcSQLException,
 :stacktrace
 ["org.h2.message.DbException.getJdbcSQLException(DbException.java:357)"
  "org.h2.message.DbException.get(DbException.java:179)"
  "org.h2.message.DbException.get(DbException.java:155)"
  "org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:150)"
  "org.h2.expression.Comparison.optimize(Comparison.java:177)"
  "org.h2.command.dml.Select.prepare(Select.java:861)"
  "org.h2.command.Parser.prepareCommand(Parser.java:283)"
  "org.h2.engine.Session.prepareLocal(Session.java:611)"
  "org.h2.engine.Session.prepareCommand(Session.java:549)"
  "org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1247)"
  "org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)"
  "org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:1135)"
  "com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:267)"
  "--> driver.sql_jdbc.execute$eval46930$fn__46931.invoke(execute.clj:239)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invokeStatic(execute.clj:257)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invoke(execute.clj:254)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:387)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
  "driver.sql_jdbc$eval59630$fn__59631.invoke(sql_jdbc.clj:49)"
  "driver.h2$eval84777$fn__84778.invoke(h2.clj:82)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
  "query_processor.context.default$default_runf.invoke(default.clj:67)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__53420.invoke(mbql_to_native.clj:26)"
  "query_processor.middleware.check_features$check_features$fn__52642.invoke(check_features.clj:42)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__53619.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__58101.invoke(wrap_value_literals.clj:137)"
  "query_processor.middleware.annotate$add_column_info$fn__47843.invoke(annotate.clj:579)"
  "query_processor.middleware.permissions$check_query_permissions$fn__52501.invoke(permissions.clj:64)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__54175.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__52731.invoke(cumulative_aggregations.clj:61)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__54749.invoke(resolve_joins.clj:183)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__50968.invoke(add_implicit_joins.clj:245)"
  "query_processor.middleware.limit$limit$fn__53398.invoke(limit.clj:38)"
  "query_processor.middleware.format_rows$format_rows$fn__53371.invoke(format_rows.clj:81)"
  "query_processor.middleware.desugar$desugar$fn__52803.invoke(desugar.clj:22)"
  "query_processor.middleware.binning$update_binning_strategy$fn__51705.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__52295.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__50505.invoke(add_dimension_projections.clj:272)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__50716.invoke(add_implicit_clauses.clj:141)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__51131.invoke(add_source_metadata.clj:105)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__54400.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__51338.invoke(auto_bucket_datetimes.clj:125)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__52350.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.parameters$substitute_parameters$fn__54151.invoke(parameters.clj:114)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__52404.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__53067.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__51146.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__57933.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54417$fn__54421.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__54417.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__53293.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__57950$fn__57951.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__57950.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__52247.invoke(cache.clj:209)"
  "query_processor.middleware.validate$validate_query$fn__57967.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__53441.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__50994.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__57912.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__52666.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__54266.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__52595.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___46655$thunk__46656.invoke(reducible.clj:101)"
  "query_processor.reducible$async_qp$qp_STAR___46655.invoke(reducible.clj:107)"
  "query_processor$base_qp$fn__58110.doInvoke(query_processor.clj:105)"
  "query_processor.reducible$sync_qp$qp_STAR___46664$fn__46667.invoke(reducible.clj:133)"
  "query_processor.reducible$sync_qp$qp_STAR___46664.invoke(reducible.clj:132)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:213)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:209)"
  "query_processor$eval58147$process_query_and_save_execution_BANG___58156$fn__58159.invoke(query_processor.clj:225)"
  "query_processor$eval58147$process_query_and_save_execution_BANG___58156.invoke(query_processor.clj:217)"
  "query_processor$eval58191$process_query_and_save_with_max_results_constraints_BANG___58200$fn__58203.invoke(query_processor.clj:237)"
  "query_processor$eval58191$process_query_and_save_with_max_results_constraints_BANG___58200.invoke(query_processor.clj:230)"
  "api.dataset$fn__62135$fn__62138.invoke(dataset.clj:54)"
  "query_processor.streaming$streaming_response_STAR_$fn__42641$fn__42642.invoke(streaming.clj:73)"
  "query_processor.streaming$streaming_response_STAR_$fn__42641.invoke(streaming.clj:72)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
  "async.streaming_response$do_f_async$fn__29624.invoke(streaming_response.clj:85)"],
 :context :ad-hoc,
 :error
 "Column \"source.Last year\" not found; SQL statement:\n-- Metabase:: userID: 1 queryType: MBQL queryHash: 11abbb927684a66d0db53bd8b9ce59a49e49d2f0f4133f29a99ee07a89f684d0\nSELECT \"source\".\"CATEGORY\" AS \"CATEGORY\", \"source\".\"avg\" AS \"avg\", \"source\".\"count-where\" AS \"count-where\" FROM (SELECT \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" AS \"CATEGORY\", avg(\"PUBLIC\".\"PRODUCTS\".\"PRICE\") AS \"avg\", sum(CASE WHEN parsedatetime(year(\"PUBLIC\".\"PRODUCTS\".\"CREATED_AT\"), 'yyyy') = parsedatetime(year(dateadd('year', -1, now())), 'yyyy') THEN 1 ELSE 0.0 END) AS \"count-where\" FROM \"PUBLIC\".\"PRODUCTS\" GROUP BY \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" ORDER BY \"PUBLIC\".\"PRODUCTS\".\"CATEGORY\" ASC) \"source\" WHERE \"source\".\"Last year\" > 0 LIMIT 2000 [42122-197]",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 4,
  :query
  {:source-query
   {:source-table 10,
    :aggregation
    [[:aggregation-options [:avg [:field-id 104]] {:name "avg"}]
     [:aggregation-options [:count-where [:= [:datetime-field [:field-id 108] :year] [:relative-datetime -1 :year]]] {:display-name "Last year", :name "count-where"}]],
    :breakout [[:field-id 107]],
    :order-by [[:asc [:field-id 107]]]},
   :filter [:> [:field-literal "Last year" :type/Float] [:value 0 nil]],
   :source-metadata
   [{:name "CATEGORY",
     :id 107,
     :table_id 10,
     :display_name "Category",
     :base_type :type/Text,
     :special_type :type/Category,
     :fingerprint {:global {:distinct-count 4, :nil% 0.0}, :type {:type/Text {:percent-json 0.0, :percent-url 0.0, :percent-email 0.0, :average-length 6.375}}},
     :settings nil}
    {:name "avg", :display_name "Average of Price", :base_type :type/Float, :special_type :type/Cost, :settings nil}
    {:name "count-where", :display_name "Last year", :base_type :type/Integer, :special_type :type/Number}],
   :fields [[:field-literal "CATEGORY" :type/Text] [:field-literal "avg" :type/Float] [:field-literal "count-where" :type/Integer]],
   :limit 2000},
  :type :query,
  :middleware {:add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [17, -85, -69, -110, 118, -124, -90, 109, 13, -75, 59, -40, -71, -50, 89, -92, -98, 73, -46, -16, -12, 19, 63, 41, -87, -98, -32, 122, -119, -10, -124, -48]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}

07-04 14:35:54 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 634.1 ms (9 DB calls) App DB connections: 0/13 Jetty threads: 3/50 (4 idle, 0 queued) (147 total active threads) Queries in flight: 1 (0 queued)
```

</details>

5. Changing the filter from the Custom Expression "Last year" and to "Average of Price", then it works:
![image](https://user-images.githubusercontent.com/1447303/86512561-9445c080-be03-11ea-8bb3-35b5aad13fb5.png)

**Information about your Metabase Installation:**
Metabase 0.35.4 and `master`

**Additional context**
Very likely the fix for #12762/#12507 will also fix this. Related to #6239
https://discourse.metabase.com/t/filter-on-countif-summarized-question/10760

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by PauloHFM** [comment]
Hi @flamber! I'm dealing with this for some time... Can't filter or order a custom expression! Seems like alias problem on generating the SQL!

Always:
ERROR: column source.'column_name_inputed' does not exist

Do you guys have this mapped in some release?

Thanks!!!

**Comment by sbelak** [comment]
The root cause is that the FE generates the wrong query, using "Last Year" (ie. `:display_name`) as the key for `:field-literal` rather than the `:name`. What I'm not sure is if the FE even knows what the `:name` is set to (ie. do we duplicate the naming logic). If not, we should find a way where we don't need to make FE overly smart.

**Comment by alekseyp** [comment]
I'm sure you guys know, but just in case - looks like #12762 didn't fix it.

**Comment by flamber** [comment]
@alekseyp Can you please open a new issue with steps-to-reproduce with Sample Dataset?

**Comment by alekseyp** [comment]
That's basically what https://github.com/metabase/metabase/issues/13857 is about

**Comment by gggdomi** [comment]
FYI, since it may be related, the exact same error occurs when using "Last Year" in "Custom Column" instead of "Filter".

(this "Custom Column" button that comes after aggregation and is absent from the original screenshot)
<img width="944" alt="Capture d’écran 2020-12-17 à 09 28 54" src="https://user-images.githubusercontent.com/11262934/102462618-4e478680-404a-11eb-8136-0ff8f0ac277d.png">

**Comment by flamber** [comment]
@gggdomi And that's #13857

**Comment by camilagarcias** [comment]
Related to this issue. It is not possible at the moment to filter over `countIf` aggregation columns

**How I made it work**
Open the SQL query to see how they named these aggregation columns. Normally, when the agg function is `countIf`the column alias are `count-where`, `count-where_2`, and so on.

Change the name of the columns to `count-where` and you'll be able to filter 


![image](https://user-images.githubusercontent.com/69139381/111980875-311dca80-8b07-11eb-97c8-235e690200ca.png)

**Comment by flamber** [comment]
Fixed by #16259
